ffi/src/primitives.rs

```diff
@@ -0,0 +1,222 @@
+    #[allow(clippy::let_unit_value)]
+    unsafe fn try_from_repr_c(source: Self::Source, _: &mut ()) -> Result<&[Self], FfiResult> {
+        let mut store = ();
+        let source: &[i8] = TryFromReprC::try_from_repr_c(source, &mut store)?;
+
+        if !source.iter().all(|e| *e == -1 || *e == 0 || *e == 1) {
+            return Err(FfiResult::TrapRepresentation);
+        }
+
+        Ok(&*(source as *const _ as *const _))
+    }
+}
+
+impl IntoFfi for core::cmp::Ordering {
+    type Target = i8;
+
+    fn into_ffi(self) -> Self::Target {
+        self as i8
+    }
+}
+impl IntoFfi for &core::cmp::Ordering {
+    type Target = *const i8;
+
+    fn into_ffi(self) -> Self::Target {
+        (self as *const core::cmp::Ordering).cast()
+    }
+}
+impl IntoFfi for &mut core::cmp::Ordering {
+    type Target = *mut i8;
+
+    fn into_ffi(self) -> Self::Target {
+        (self as *mut core::cmp::Ordering).cast()
+    }
+}
+
+impl<'itm> IntoFfiSliceRef<'itm> for core::cmp::Ordering {
+    type Target = SliceRef<'itm, i8>;
+
+    fn into_ffi(source: &[Self]) -> Self::Target {
+        // SAFETY: `core::cmp::Ordering` has the same representation as i8
+        unsafe { SliceRef::from_slice(&*(source as *const [_] as *const [i8])) }
+    }
+}
+
+macro_rules! primitive_impls {
+    ( $( $ty:ty ),+ $(,)? ) => { $(
+        unsafe impl ReprC for $ty {}
+
+        impl TryFromReprC<'_> for $ty {
+            type Source = Self;
+            type Store = ();
+
+            unsafe fn try_from_repr_c(source: Self::Source, _: &mut Self::Store) -> Result<Self, FfiResult> {
+                Ok(source)
+            }
+        }
+
+        impl<'itm> TryFromReprCSliceRef<'itm> for $ty {
+            type Source = SliceRef<'itm, $ty>;
+            type Store = ();
+
+            unsafe fn try_from_repr_c(source: Self::Source, _: &'itm mut Self::Store) -> Result<&[Self], FfiResult> {
+                source.into_slice().ok_or(FfiResult::ArgIsNull)
+            }
+        }
+
+        impl<'slice> TryFromReprCSliceMut<'slice> for $ty {
+            type Source = SliceMut<'slice, $ty>;
+            type Store = ();
+
+            unsafe fn try_from_repr_c(source: Self::Source, _: &mut Self::Store) -> Result<&'slice mut [Self], FfiResult> {
+                source.into_slice().ok_or(FfiResult::ArgIsNull)
+            }
+        }
+
+        impl IntoFfi for $ty {
+            type Target = Self;
+
+            fn into_ffi(self) -> Self::Target {
+                self
+            }
+        }
+
+        impl<'itm> IntoFfiSliceRef<'itm> for $ty {
+            type Target = SliceRef<'itm, $ty>;
+
+            fn into_ffi(source: &[Self]) -> Self::Target {
+                SliceRef::from_slice(source)
+            }
+        }
+        unsafe impl<'slice> IntoFfiSliceMut<'slice> for $ty {
+            type Target = SliceMut<'slice, $ty>;
+
+            fn into_ffi(source: &mut [Self]) -> Self::Target {
+                SliceMut::from_slice(source)
```

Comment:
Are you sure that it is ok for `f32` and `f64`? they are non-robust as i understend.

Reply:
Good point, I'll check

Reply:
https://app.zenhub.com/workspaces/iroha-v2-60ddb820813b9100181fc060/issues/hyperledger/iroha/2512